include/seastar/core/semaphore.hh

```diff
@@ -165,37 +166,94 @@ private:
         state(basic_semaphore& sem_, ssize_t count_) noexcept : sem(&sem_), count(count_) {}
     };
     lw_shared_ptr<state> _state;
-    struct entry {
-        promise<> pr;
+
+    class basic_waiter : public boost::intrusive::list_base_hook<boost::intrusive::link_mode<boost::intrusive::auto_unlink>> {
+    public:
+        basic_semaphore* sem;
         size_t nr;
-        std::optional<abort_on_expiry<clock>> timer;
-        entry(promise<>&& pr_, size_t nr_) noexcept : pr(std::move(pr_)), nr(nr_) {}
-    };
-    struct expiry_handler {
-        basic_semaphore& sem;
-        void operator()(entry& e) noexcept {
-            if (e.timer) {
+
+    private:
+        std::variant<optimized_optional<abort_source::subscription>, timer> _aborter;
```

Comment:
Last time I tried to use the variant for exactly same purposes I got the comment that it is expensive for such a low level primitive:

https://groups.google.com/g/seastar-dev/c/go80xYaW_tw/m/zaA22muuAwAJ

Reply:
if we don't use `std::visit<>` to perform dispatching, probably we don't need to worry about the expensive operations? as `std::variant<>` is but an recursive union. if we just use it for a type safe storage, we should be fine, i think.

regarding to the memory footprint,  it uses extra `sizeof(char)` bytes for storing the index / discriminator, as it's large enough for representing 0 and 1 . as the `sizeof(timer)` is 96, and `sizeof(abort_source::subscription)` is 56, the `sizeof(variant<...>)` is 104, taking the padding into consideration. if we always know the exact type stored in the variant, and we care about the extra memory space used for the index, surely we should use a more performant container for holding these two types. but we still need a discriminator, so we can call the right destructor.

Reply:
> if we don't use `std::visit<>` to perform dispatching, probably we don't need to worry about the expensive operations? as `std::variant<>` is but an recursive union. if we just use it for a type safe storage, we should be fine, i think.
> 
Made the same argument in the thread...

Reply:
i see. Avi didn't like the feature-rich interfaces exposed by `std::variant<>` and friends, and only a tiny subset of which is used on such a low-level structure. it's like purchasing a spaceship for commute. but the same applies to the operation system, editor and the C++ language we are using.  as we only pay for what we use. that's also an important design principle of C++.

anyhow, it's might not be relevant anymore. was just curious why.

Reply:
> if we just use it for a type safe storage, we should be fine, i think.
agreed.

> we still need a discriminator, so we can call the right destructor.
otherwise, a plain union could have sufficient for our needs.

I'll take a stab at using object-oriented inheritance instead so a timeout_waiter and abortable_waiter could be derived from the non-abortable basic_waiter.

Reply:
inheritance was too hairy as the bases class is used by units_waiter with the addition of atomic get_units methods
and both wait and get_units need all three "dialects": plain, timeout, and abort_source.